ensure meta impls for nonfunctional collectives are registered on import torch; support nonfunctional collectives with FakeTensor

bdhirsh opened the pull request:
non-functional collectives (e.g. `torch.distributed.all_gather`) will work with FakeTensorMode automatically as long as each underlying dispatcher operator has a `Meta` implementation.

These ops already have `Meta` impls, but previously they would only get registered if you `import torch.distributed._tools.fake_collectives`.

This PR tries to ensure that they get registered at `import torch` time (if distributed is available), around the same place that we register the python meta impls of our aten ops.

local test:
```
import torch
import torch.distributed as dist
from torch.distributed.tensor import DeviceMesh
from torch.testing._internal.distributed.fake_pg import FakeStore
from torch.distributed.tensor import DeviceMesh, DTensor, Replicate

device_type = "cuda"
world_size = 4

fake_store = FakeStore()
dist.init_process_group(
    "fake", store=fake_store, rank=0, world_size=world_size
)
default_pg = torch.distributed.distributed_c10d._get_default_group()

import torch.distributed._functional_collectives as funcol
from torch._subclasses import FakeTensorMode
with FakeTensorMode():
    inp = torch.randn(4, 4, device=device_type)

    dist.all_reduce(inp, group=default_pg)
```

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #171196
* __->__ #162119



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @ezyang

bdhirsh committed:
ensure meta impls for nonfunctional collectives are registered on import torch; support nonfunctional collectives with FakeTensor

[ghstack-poisoned]

ezyang commented on torch/__init__.py:
```diff
@@ -2667,6 +2667,12 @@ def _register_device_module(device_type, module):
 if not TYPE_CHECKING:
     from torch import _meta_registrations
 
+    if torch.distributed.is_available():
```
BTW this is always True now post https://github.com/pytorch/pytorch/pull/160449

ezyang commented:
I'm having trouble understanding how this can work, because in an unrelated pdb step through:

```
-> dist.all_reduce(lt_sum, group=fake_pg)
(Pdb) s
--Call--
> /Users/ezyang/Dev/pytorch/torch/distributed/c10d_logger.py(78)wrapper()
-> @functools.wraps(func)
(Pdb) n
> /Users/ezyang/Dev/pytorch/torch/distributed/c10d_logger.py(80)wrapper()
-> try:
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/c10d_logger.py(81)wrapper()
-> return func(*args, **kwargs)
(Pdb) s
--Call--
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2828)all_reduce()
-> @_exception_logger
(Pdb) n
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2882)all_reduce()
-> relevant_args = (tensor,)
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2883)all_reduce()
-> if has_torch_function(relevant_args):
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2893)all_reduce()
-> _check_single_tensor(tensor, "tensor")
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2894)all_reduce()
-> if _rank_not_in_group(group):
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2898)all_reduce()
-> if tensor.is_complex():
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2903)all_reduce()
-> opts = AllreduceOptions()
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2904)all_reduce()
-> opts.reduceOp = op
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2905)all_reduce()
-> opts.asyncOp = async_op
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2906)all_reduce()
-> if group is None:
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2909)all_reduce()
-> if group in _world.pg_coalesce_state.keys():
(Pdb)
> /Users/ezyang/Dev/pytorch/torch/distributed/distributed_c10d.py(2918)all_reduce()
-> work = group.allreduce([tensor], opts)
```

ezyang commented:
Redispatch will happen if you get to the superclass of the PG:

```
  virtual c10::intrusive_ptr<Work> allreduce(
      std::vector<at::Tensor>& tensors,
      const AllreduceOptions& opts = AllreduceOptions()) {
    static auto op =
        c10::Dispatcher::singleton()
            .findSchemaOrThrow("c10d::allreduce_", "")
            .typed<
                std::tuple<std::vector<at::Tensor>, c10::intrusive_ptr<Work>>(
                    at::TensorList,
                    const c10::intrusive_ptr<::c10d::ProcessGroup>&,
                    const c10::intrusive_ptr<::c10d::ReduceOp>&,
                    const std::optional<at::Tensor>& sparse_indices,
                    bool,
                    int64_t)>();

    auto work = std::get<1>(op.call(
        tensors,
        c10::intrusive_ptr<ProcessGroup>::unsafe_reclaim_from_nonowning(this),
        c10::make_intrusive<ReduceOp>(opts.reduceOp),
        opts.sparseIndices,
        opts.asyncOp,
        opts.timeout.count()));

    if (c10d::allow_inflight_collective_as_graph_input()) {
      for (const auto& tensor : tensors) {
        c10d::register_work(tensor, work);
      }
    }
    return work;
  }
```

bdhirsh committed:
Update on "ensure meta impls for nonfunctional collectives are registered on import torch; support nonfunctional collectives with FakeTensor"

non-functional collectives (e.g. `torch.distributed.all_gather`) will work with FakeTensorMode automatically as long as each underlying dispatcher operator has a `Meta` implementation.

These ops already have `Meta` impls, but previously they would only get registered if you `import torch.distributed._tools.fake_collectives`.

This PR tries to ensure that they get registered at `import torch` time (if distributed is available), around the same place that we register the python meta impls of our aten ops.

local test:
```
import torch
import torch.distributed as dist
from torch.distributed.tensor import DeviceMesh
from torch.testing._internal.distributed.fake_pg import FakeStore
from torch.distributed.tensor import DeviceMesh, DTensor, Replicate

device_type = "cuda"
world_size = 4

fake_store = FakeStore()
dist.init_process_group(
    "fake", store=fake_store, rank=0, world_size=world_size
)
default_pg = torch.distributed.distributed_c10d._get_default_group()

import torch.distributed._functional_collectives as funcol
from torch._subclasses import FakeTensorMode
with FakeTensorMode():
    inp = torch.randn(4, 4, device=device_type)

    dist.all_reduce(inp, group=default_pg)
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge -i

bdhirsh committed:
Update on "ensure meta impls for nonfunctional collectives are registered on import torch; support nonfunctional collectives with FakeTensor"

non-functional collectives (e.g. `torch.distributed.all_gather`) will work with FakeTensorMode automatically as long as each underlying dispatcher operator has a `Meta` implementation.

These ops already have `Meta` impls, but previously they would only get registered if you `import torch.distributed._tools.fake_collectives`.

This PR tries to ensure that they get registered at `import torch` time (if distributed is available), around the same place that we register the python meta impls of our aten ops.

local test:
```
import torch
import torch.distributed as dist
from torch.distributed.tensor import DeviceMesh
from torch.testing._internal.distributed.fake_pg import FakeStore
from torch.distributed.tensor import DeviceMesh, DTensor, Replicate

device_type = "cuda"
world_size = 4

fake_store = FakeStore()
dist.init_process_group(
    "fake", store=fake_store, rank=0, world_size=world_size
)
default_pg = torch.distributed.distributed_c10d._get_default_group()

import torch.distributed._functional_collectives as funcol
from torch._subclasses import FakeTensorMode
with FakeTensorMode():
    inp = torch.randn(4, 4, device=device_type)

    dist.all_reduce(inp, group=default_pg)
```




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot rebase -b main

bobrenjc93 committed:
Update

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 committed:
Update

[ghstack-poisoned]

bobrenjc93 committed:
Update

[ghstack-poisoned]

bobrenjc93 committed:
Update

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge
